src/dawn/Optimizer/PassSetNonTempCaches.cpp

```diff
@@ -0,0 +1,325 @@
+    domethod->getStatementAccessesPairs().clear();
+
+    for(int i = 0; i < assignmentIDs.size(); ++i) {
+      int assignmentID = assignmentIDs[i];
+      int assigneeID = assigneeIDs[i];
+      addAssignmentToDoMethod(domethod, assignmentID, assigneeID);
+    }
+
+    // Add the single do method to the new Stage
+    assignmentStage->getDoMethods().clear();
+    assignmentStage->addDoMethod(domethod);
+    assignmentStage->update();
+
+    return assignmentStage;
+  }
+
+  ///@brief add the assignment operator of two unique id's to a given domethod
+  void addAssignmentToDoMethod(std::unique_ptr<DoMethod>& domethod, int assignmentID,
+                               int assigneeID) {
+    // Create the StatementAccessPair of the assignment with the new and old variables
+    std::shared_ptr<FieldAccessExpr> fa_assignee =
+        std::make_shared<FieldAccessExpr>(instantiation_->getNameFromAccessID(assigneeID));
+    std::shared_ptr<FieldAccessExpr> fa_assignment =
+        std::make_shared<FieldAccessExpr>(instantiation_->getNameFromAccessID(assignmentID));
+    std::shared_ptr<AssignmentExpr> assignmentExpression =
+        std::make_shared<AssignmentExpr>(fa_assignment, fa_assignee, "=");
+    std::shared_ptr<ExprStmt> expAssignment = std::make_shared<ExprStmt>(assignmentExpression);
+    std::shared_ptr<Statement> assignmentStatement =
+        std::make_shared<Statement>(expAssignment, nullptr);
+    std::shared_ptr<StatementAccessesPair> pair =
+        std::make_shared<StatementAccessesPair>(assignmentStatement);
+    std::shared_ptr<Accesses> newAccess = std::make_shared<Accesses>();
+    newAccess->addWriteExtent(assignmentID, Extents({0, 0, 0}));
+    newAccess->addReadExtent(assigneeID, Extents({0, 0, 0}));
+    pair->setAccesses(newAccess);
+    domethod->getStatementAccessesPairs().push_back(pair);
+
+    // Add the new expressions to the map
+    instantiation_->getExprToAccessIDMap().emplace(fa_assignment, assignmentID);
+    instantiation_->getExprToAccessIDMap().emplace(fa_assignee, assigneeID);
+  }
+
+  /// @brief checks if there is a read operation before the first write operation in the given
+  /// multistage.
+  /// @param[in]    id  original FieldAccessID of the field to check
+  /// @return           true if there is a read-access before the first write access
+  bool checkReadBeforeWrite(int id) {
+    for(auto stageItGlob = multiStagePrt_->getStages().begin();
+        stageItGlob != multiStagePrt_->getStages().end(); ++stageItGlob) {
+      DoMethod& doMethod = (**stageItGlob).getSingleDoMethod();
+      for(int stmtIndex = 0; stmtIndex < doMethod.getStatementAccessesPairs().size(); ++stmtIndex) {
+        const std::shared_ptr<StatementAccessesPair>& stmtAccessesPair =
+            doMethod.getStatementAccessesPairs()[stmtIndex];
+
+        // Find first if this statement has a read
+        auto readAccessIterator = stmtAccessesPair->getCallerAccesses()->getReadAccesses().find(id);
```

Comment:
It think for sake of readability it is worth storing the read accesses in a local variable to avoid the long chains `stmtAccessesPair->getCallerAccesses()->getReadAccesses()` (you access it several times)

Reply:
it actually is only 2 consecutive calls. Is it really worth it for that?